### ISSUE TITLE
Expose $el while initializing x-data

### ISSUE THREAD

**Issue by jonaskuske** [alpinejs/alpine#471]
Would be really cool if `$el` was already available inside `x-data`, so we can set initial state that depends on the DOM element. My use case is fading in an image, where I need to set the initial state so that it reflects whether the image already finished loading before Alpine was ready or not:

```html
<img
  x-data="{ loaded: false }"
  x-init="loaded = Boolean($el.complete && $el.naturalWidth && $el.naturalHeight)"
  @load="loaded = true"
/>
```

Would be cool if I could skip the `x-init` here and directly initialize `x-data` to the correct value instead of setting a placeholder. So in the best case I could do:

```html
<img x-data="{ loaded: isImageLoaded($el) }" @load="loaded = true" />
```

**Comment by HugoDF** [comment]
I'm not sure it makes sense for `$el` to be available to `x-data`.

If we look at it in terms of architecture/spirit:

- `x-data` expects a (mostly) side-effect free/in-memory/synchronous "create an object" type of function/inline object. There's no reason for it to know about the element since it should just return some data to initialise the component with.
- `x-init` is the mount/unmount hook definition. It runs after component initialisation and it's expected that this is where you access/modify the DOM, fetch data from APIs etc.

Which is why it makes sense (in spirit) for `x-init` to have access to `$el` but not `x-data` (although I'm sure we could actually make it have access to `$el` since `x-data` is on the `$el`)

Let's see what @calebporzio thinks.

**Comment by jonaskuske** [comment]
> * There's no reason for it to know about the element since it should just return some data to initialise the component with.

...unless that data depends on the element. But that's only ever relevant for elements that have their own lifecycle (images, autoplaying videos...) and I 100% understand if that usecase is too niche. 😅
Especially as setting some arbitrary value as placeholder and then using `x-init` to get the actual data object is ok as a workaround.


> * `x-init` is the mount/unmount hook definition. It runs after component initialisation and it's expected that this is where you access/modify the DOM, fetch data from APIs etc.
> 
> Which is why it makes sense (in spirit) for `x-init` to have access to `$el` but not `x-data` (although I'm sure we could actually make it have access to `$el` since `x-data` is on the `$el`)

Well they're not 1:1 anyway. Alpine already gives you access to `$el` before initialization is done and DOM updates are applied, which Vue doesn't and can't. (in Vue, `created` → no `$el` and `mounted` → `$el`, with all updates applied already)
`x-init` is basically *"Vue's `created` but we give you `$el` already"*, so you could see `$el` being available in `x-data` as staying true to the spirit of "we're an HTML-first framework and already have the DOM element anyway, so might as well let you access it earlier that other frameworks can". :)

But as I said, 100% understandable if you don't want to do this, it's only a small convenience thing anyway. But since from what I understand Alpine generally seems to favor convenience over an React-ish "pure no matter the cost" approach so I thought I'd ask 👍

Still, mentioning in the docs that the "magic properties" are not available yet in `x-data` is probably a good idea.

**Comment by HugoDF** [comment]
> But as I said, 100% understandable if you don't want to do this, it's only a small convenience thing anyway. But since from what I understand Alpine generally seems to favor convenience over an React-ish "pure no matter the cost" approach I thought I'd ask 👍
>
> Still, mentioning in the docs that the "magic properties" are not available yet in x-data is probably a good idea.

You're right, let's see what Caleb thinks of exposing `$el` in `x-data`, you can create a PR that does it, he seems reasonably happy to merge stuff that doesn't add too much bulk to Alpine.

Otherwise yeah a note in the README is a good shout 😄

**Comment by devcircus** [comment]
Yeah this doesn't seem possible without complicating the current initialization process quite a bit. Thinking a note in the docs is the way to go.
